File: TheaterSuggestions/Java/SeatsSuggestions/src/main/java/com/baasie/SeatsSuggestions/SuggestionsMade.java
package com.baasie.SeatsSuggestions;

import java.util.*;
import java.util.stream.Collectors;

public class SuggestionsMade {
    private final String showId;
    private final int partyRequested;

    public Map<PricingCategory, List<SuggestionMade>> forCategory = new HashMap<>();

    public SuggestionsMade(String showId, int partyRequested) {
        this.showId = showId;
        this.partyRequested = partyRequested;

        instantiateAnEmptyListForEveryPricingCategory();
    }

    public List<String> seatNames(PricingCategory pricingCategory) {
        return forCategory.get(pricingCategory).stream().map(s -> String.join("-", s.seatNames())).collect(Collectors.toList());
    }

    private void instantiateAnEmptyListForEveryPricingCategory() {
        for (PricingCategory pricingCategory : PricingCategory.values()) {
            forCategory.put(pricingCategory, new ArrayList<>());
        }
    }

    public void add(Iterable<SuggestionMade> suggestions) {
        suggestions.forEach(suggestionMade -> forCategory.get(suggestionMade.pricingCategory()).add(suggestionMade));
    }

    public boolean matchExpectations() {
        return forCategory.values().stream().flatMap(List::stream).anyMatch(SuggestionMade::MatchExpectation);
    }
}

File: TheaterSuggestions/CSharp/SeatsSuggestions.Domain/SuggestionsMade.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Value;
using Value.Shared;

namespace SeatsSuggestions.Domain
{
    /// <summary>
    ///     Occurs when a bunch of Suggestion are made.
    /// </summary>
    public class SuggestionsMade: ValueType<SuggestionsMade>, IProvideDomainEvent
    {
        public ShowId ShowId { get; }
        public PartyRequested PartyRequested { get; }

        private Dictionary<PricingCategory, ListByValue<SuggestionMade>> ForCategory { get; } =
            new Dictionary<PricingCategory, ListByValue<SuggestionMade>>();

        public IEnumerable<string> SeatsInFirstPricingCategory => SeatNames(PricingCategory.First);
        public IEnumerable<string> SeatsInSecondPricingCategory => SeatNames(PricingCategory.Second);
        public IEnumerable<string> SeatsInThirdPricingCategory => SeatNames(PricingCategory.Third);
        public IEnumerable<string> SeatsInMixedPricingCategory => SeatNames(PricingCategory.Mixed);

        public SuggestionsMade(ShowId showId, PartyRequested partyRequested)
        {
            ShowId = showId;
            PartyRequested = partyRequested;
            InstantiateAnEmptyListForEveryPricingCategory();
        }

        public IEnumerable<string> SeatNames(PricingCategory pricingCategory)
        {
            var suggestionsMade = ForCategory[pricingCategory];
            return suggestionsMade.Select(s => string.Join("-", s.SeatNames()));
        }

        private void InstantiateAnEmptyListForEveryPricingCategory()
        {
            foreach (PricingCategory pricingCategory in Enum.GetValues(typeof(PricingCategory)))
                ForCategory[pricingCategory] = new ListByValue<SuggestionMade>();
        }

        public void Add(IEnumerable<SuggestionMade> suggestions)
        {
            foreach (var suggestionMade in suggestions) ForCategory[suggestionMade.PricingCategory].Add(suggestionMade);
        }

        public bool MatchExpectations()
        {
            return ForCategory.SelectMany(s => s.Value).Any(x => x.MatchExpectation());
        }

        protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
        {
            return new object[] { ShowId, PartyRequested, new DictionaryByValue<PricingCategory, ListByValue<SuggestionMade>>(ForCategory) };
        }
    }
}